src/Microsoft.Framework.Runtime.Hosting/Json/JsonBuffer.cs

```diff
@@ -0,0 +1,301 @@
+            }
+            else if (first == JsonConstants.Comma)
+            {
+                result.Type = JsonTokenType.Comma;
+            }
+            else if (first == JsonConstants.Quotation)
+            {
+                result.Type = JsonTokenType.String;
+                result.Value = ReadString();
+            }
+            else if (first == (int)'t')
+            {
+                ReadLiteral(JsonConstants.ValueTrue);
+                result.Type = JsonTokenType.True;
+            }
+            else if (first == (int)'f')
+            {
+                ReadLiteral(JsonConstants.ValueFalse);
+                result.Type = JsonTokenType.False;
+            }
+            else if (first == (int)'n')
+            {
+                ReadLiteral(JsonConstants.ValueNull);
+                result.Type = JsonTokenType.Null;
+            }
+            else if (char.IsDigit((char)first) || first == (int)'-')
+            {
+                result.Type = JsonTokenType.Number;
+                result.Value = ReadNumber(first);
+            }
+            else
+            {
+                throw new JsonDeserializerException(string.Format("Illegal character {0:X4}.", first), result.GetPosition());
+            }
+
+            return result;
+        }
+
+        private int ReadNextChar()
+        {
+            while (true)
+            {
+                var value = _reader.Read();
+                _column += 1;
+
+                if (value == -1)
+                {
+                    // This is the end of file
+                    return -1;
+                }
+                else if (value == JsonConstants.LF)
+                {
+                    // This is a new line. Let the next loop read the first charactor of the following line.
+                    // Set position ahead of next line
+                    _column = 0;
+                    _line += 1;
+
+                    continue;
+                }
+                else if (value == JsonConstants.CR)
+                {
+                    // Skip the carriage return.
+                    // Let the next loop read the following char
+                }
+                else
+                {
+                    // Returns the normal value
+                    return value;
+                }
+            }
+        }
+
+        private string ReadNumber(int firstRead)
+        {
+            var buf = new List<char> { (char)firstRead };
+            while (true)
+            {
+                var next = _reader.Peek();
+
+                if (char.IsDigit((char)next) ||
+                    next == JsonConstants.FullStop ||
+                    next == JsonConstants.UpperE ||
+                    next == JsonConstants.LowerE)
+                {
+                    buf.Add((char)ReadNextChar());
+                }
+                else
+                {
+                    break;
+                }
+            }
+
+            return new string(buf.ToArray());
```

Comment:
I think I saw a comment from @victorhurdugaci or something to just use StringBuilder directly. With the code here there are *three* copies of the data: 1. `List<char>`, 2. `char[]`, 3. `string`.

I recommend either using StringBuilder (which will have less copying/allocations), or find a way to use the `List<char>` or whatever directly (if possible).